keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1149 @@
+where an object containing an unrecognized field was transcoded from YAML or
+JSON to CBOR. A strict decode error (as in JSON) may be appropriate in this case.
+
+This is not a concern for the custom resource path, which uses a schema-aware
+decoder wrapper to detect unknown fields.
+<<[/UNRESOLVED]>>
+
+Note that clients (e.g. kubectl) may choose to decode an object from a JSON or
+YAML text representation containing duplicate keys, then encode to CBOR to
+populate the body of an API request. Since the text-encoded content (potentially
+containing duplicates) is not literally transcoded to CBOR, this use case is
+supported. Depending on the strictness mode, duplicate keys would either be
+removed or produce an error at decode time.
+
+### Encoding Determinism
+
+It is possible for a single object to be encodable as multiple distinct but
+valid and semantically-equivalent CBOR byte strings. The CBOR specification does
+not require encoder implementations to produce deterministic output, although it
+does include recommendations for implementing deterministic encoding.
+
+The etcd3 storage implementation of GuaranteedUpdate relies on deterministic
+encoding to skip writes if the stored bytes would not change. The existing JSON
+and Protobuf encoders produce deterministic output.
+
+Encode benchmarks for the two evaluated Go CBOR libraries show a 2.4x speedup
+and a 1.8x speedup by disabling map key sorting. According to the spec, “the
+CBOR data model for maps does not allow ascribing semantics to the order of the
+key/value pairs in the map representation.”  And since the CBOR decoder will
+reject maps containing duplicate keys, a CBOR map represents exactly the same
+set of key-value pairs regardless of the order they are encoded.
+
+In order to take advantage of the available speedup, the CBOR encoder will
+support separate deterministic and nondeterministic modes. The deterministic
+mode will be used for storage serialization only. The nondeterministic mode
+should actively randomize the order of map item encoding (as with map iteration
+in Go) to prevent users from making assumptions about the order.
+
+Clients may imagine other use cases for deterministic encoding of response
+bodies:
+
+- caching
+  - The existing WithCacheControl filter sets the response header
+  “Cache-Control: no-cache, private” to prevent shared caches from storing
+  responses (since requests are subject to authn/authz), and to prevent
+  responses in non-shared caches from being reused without
+  validation. Deterministic encoding could allow an API server to generate
+  strong ETags by hashing the encoded form of the resource.
+  - Even for the existing data formats, there should be no caching proxies
+  storing API responses.
+- diffing
+  - The human-readable text formats (JSON and YAML) are not changing under this
+  proposal.
+  - The documentation will explicitly state that this should not be done with
+  CBOR.
```

Comment:
I agree with the rationale.. but this does have the potential to be a landmine. Can we do better than documentation, are there code routines that we can encapsulate that do tasks where this is most likely to be a problem and then make it super clear in those code routines that we can't expect deterministic hashes of CBOR encoded data?

Reply:
Hm, what do you think about a method `EncodeNondeterministic(runtime.Object, io.Writer) error` either added directly to the Encoder interface, or in a separate interface optionally implemented by serializers / type-asserted in the response path? It's a little messy, but at least callers wouldn't need to break the Encoder abstraction to make different assumptions depending on the implementation. I imagine it'd also be harder for callers to ignore or miss. Of course, the farther you get from the point of encode, the less visible it would be.

Reply:
Yeah, something like this seems reasonable. Also, if we have anything that does do serialized comparisons via hashing, try to isolate that into a routine and make sure it has a special case for CBOR? Then during review if we ever see someone doing that sort of comparison, we could funnel them toward that routine?  Maybe do a quick sweep of the watch cache code to see if we have that sort of thing?

Reply:
I did a sweep over anything using `bytes.Equal` or importing/using any of the stdlib `hash/*` packages and didn't turn up anything new. There are some bytewise comparisons on `runtime.RawExtension` data. They look fine, since the responsibility for encoding and decoding falls on the same controller. It could be useful to wrap this in a function that reports encoding mismatches. Do you think that would be worthwhile?